### PR TITLE
Feat/Persistent session storage

### LINKED ISSUE #1189: Support persistent sessions
In current implementation node does not store sessions persistently. This results in the loss of open sessions when the node process is stopped. According to NeoFS API protocol user specifies session lifetime only. IMO client shouldn't worry about node process, but only about specfieid lifetime of the sessions. Persistent local session storage will minimize the impact of the process state on the session.

### PR DESCRIPTION
**Pull request by carpawell**
Closes #1189. Not sure if node admin should be able to configure encryption.

Please, double-check the encryption approach.

### CODE REVIEW

File: pkg/services/session/storage/persistent/storage.go [hunk 1]
```diff
@@ -54,7 +62,38 @@ func NewTokenStore(path string, opts ...Option) (*TokenStore, error) {
...
+
+		// tagOffset is a constant offset for
+		// tags when marshalling ECDSA key in
+		// ASN.1 DER form
+		const tagOffset = 7
+
+		// using first 32 bytes from
+		// the marshalled private key
+		// as a secret
+		c, err := aes.NewCipher(rawKey[tagOffset : tagOffset+32])
```
Comment (fyrchik) [inline pkg/services/session/storage/persistent/storage.go:83]: https://github.com/golang/go/blob/master/src/crypto/x509/sec1.go#L58
```
rawKey := make([]byte, (key.Curve.Params().N.BitLen()+7)/8)
cfg.privateKey.D.FillBytes(rawKey)
```

Also works for every private key type, not just 32-byte one.

File: pkg/services/session/storage/persistent/util.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/nspcc-dev/neofs-node/pkg/services/session/storage"
+	"go.etcd.io/bbolt"
+)
+
+const expOffset = 8
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/session/storage/persistent/util.go]: my bad, renamed

File: pkg/services/session/storage/persistent/util.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	return binary.LittleEndian.Uint64(rawToken[:expOffset])
+}
+
+func iterateNestedBuckets(b *bbolt.Bucket, fn func(b *bbolt.Bucket) error) error {
+	c := b.Cursor()
+
+	for k, v := c.First(); k != nil; k, v = c.Next() {
+		// nil value is a hallmark
+		// of the nested buckets
+		if v == nil {
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/session/storage/persistent/util.go:67]: in fact, no, but the idea that a node could potentially use storage that has not been created by it haunts me

File: pkg/services/session/storage/persistent/util.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	key, err := x509.ParseECPrivateKey(raw[expOffset:])
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal private key: %w", err)
+	}
+
+	return storage.NewPrivateToken(key, epoch), nil
+}
+
+func epochFromToken(rawToken []byte) uint64 {
+	return binary.LittleEndian.Uint64(rawToken[:expOffset])
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/session/storage/persistent/util.go]: wanted to make it more explicit, but agree that it is useless, deleted

File: pkg/services/session/storage/persistent/executor_test.go [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+			require.Nil(t, token)
+		} else {
+			equalKeys(t, test.key, token.SessionKey())
+		}
+	}
+}
+
+// This test was added to fix bolt's behaviour since the persistent
+// storage uses cursor and there is an issue about `cursor.Delete`
+// method: https://github.com/etcd-io/bbolt/issues/146.
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/session/storage/persistent/executor_test.go:159]: this is an old issue (origins to the original archived repo) and I could reproduce key skipping when did such test in the one tx (updating and iterating in one `Update` call), so technically it is still an issue (at least there is no detailed behavior description about such use-cases)

however, code written in that PR works fine, but I've decided to add that test to know when bbolt's maintainers have done something breaking about that issue

File: pkg/services/session/storage/persistent/storage.go [hunk 6]
```diff
@@ -18,6 +21,11 @@ type TokenStore struct {
 	db *bbolt.DB
 
 	l *zap.Logger
+
+	// optional AES-256 algorithm
+	// encryption in Galois/Counter
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/session/storage/persistent/storage.go:24]: got naming from [wiki](https://en.wikipedia.org/wiki/Galois/Counter_Mode)

still needed to be changed?

File: pkg/services/session/storage/persistent/storage.go (hunk 1 above)
Comment (carpawell) [reply pkg/services/session/storage/persistent/storage.go]: looks better, does the same, thanks

File: pkg/services/session/storage/persistent/util.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+func packToken(exp uint64, key *ecdsa.PrivateKey) ([]byte, error) {
+	rawKey, err := x509.MarshalECPrivateKey(key)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal private key: %w", err)
+	}
+
+	res := make([]byte, expOffset, expOffset+len(rawKey))
+	binary.LittleEndian.PutUint64(res, exp)
+
+	res = append(res, rawKey...)
```
[the remark this answers is not in the document]
Comment (carpawell) [reply pkg/services/session/storage/persistent/util.go:31]:
```go
func BenchmarkAppend(b *testing.B) {
	for i := 0; i < b.N; i++ {
		existing := make([]byte, 1000, 1000)

		init := make([]byte, 0, 1000)
		_ = append(init, existing...)
	}
}

func BenchmarkCopy(b *testing.B) {
	for i := 0; i < b.N; i++ {
		existing := make([]byte, 1000, 1000)

		init := make([]byte, 1000, 1000)
		copy(init, existing)
	}
}
```

run that, got result:

```
goos: linux
goarch: amd64
cpu: Intel(R) Core(TM) i7-10510U CPU @ 1.80GHz
BenchmarkAppend-8       79206925                14.93 ns/op
BenchmarkCopy-8         77391738                14.94 ns/op
```

File: pkg/services/session/storage/persistent/encryption_test.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+func TestTokenStore_Encryption(t *testing.T) {
+	pk, err := keys.NewPrivateKey()
+	require.NoError(t, err)
+
+	ts, err := NewTokenStore(filepath.Join(t.TempDir(), ".storage"), WithEncryptionKey(&pk.PrivateKey))
+	require.NoError(t, err)
+
+	data := []byte("nice encryption, awesome tests")
+
+	encryptedData, err := ts.encrypt(data)
```
Comment (alexvanin) [inline pkg/services/session/storage/persistent/encryption_test.go:20]: Compare that encrypted and decrypted data are not the same.

**Review by alexvanin — APPROVED** [review]

File: pkg/services/session/storage/persistent/encryption_test.go (hunk 8 above)
Comment (carpawell) [reply pkg/services/session/storage/persistent/encryption_test.go:20]: sure

**Pull request merged**